File: VB/DiagnosticTool/DignosticController/DiagnosticController.vb
Imports System
Imports System.Collections.Generic
Imports System.Diagnostics
Imports System.IO
Imports System.Linq
Imports System.Threading
Imports DashboardDiagnosticTool.Data
Imports DiagnosticTool.LinuxTools
Imports Microsoft.Diagnostics.Tracing
Imports Microsoft.Diagnostics.Tracing.Session

Namespace DashboardDiagnosticTool

    Public Class DiagnosticController
        Implements IDisposable

        Private fileName As String = ""

        Private runningSession As Integer = 0

        Private path As String

        Private OnStop As Action(Of TraceDataProcessor)

        Private ReadOnly fileController As IFileController

        Private ReadOnly place As String = "Session"

        Private ReadOnly telemetry As String = "DX-Dashboard-Telemetry"

        Private ReadOnly traceItemsField As Dictionary(Of Integer, Dictionary(Of Integer, List(Of TraceItem))) = New Dictionary(Of Integer, Dictionary(Of Integer, List(Of TraceItem)))()

        Private ReadOnly benchmarksField As Dictionary(Of Integer, List(Of BenchmarkItem)) = New Dictionary(Of Integer, List(Of BenchmarkItem))()

        Private ReadOnly sessionsField As List(Of SessionItem) = New List(Of SessionItem)()

        Public InProcess As Boolean = False

        Public ReadOnly Property FileExists As Boolean
            Get
                Return File.Exists(fileName)
            End Get
        End Property

        Public ReadOnly Property TraceItems As Dictionary(Of Integer, Dictionary(Of Integer, List(Of TraceItem)))
            Get
                Return traceItemsField
            End Get
        End Property

        Public ReadOnly Property Benchmarks As Dictionary(Of Integer, List(Of BenchmarkItem))
            Get
                Return benchmarksField
            End Get
        End Property

        Public ReadOnly Property Sessions As List(Of SessionItem)
            Get
                Return sessionsField
            End Get
        End Property

        Public Event ThrowException As Action(Of Exception)

        Public Event SessionsChanged As Action

        Public Event StartProcessing As Action

        Public Event EndProcessing As Action

        Public Sub New(ByVal fileController As IFileController)
            Me.fileController = fileController
        End Sub

        Public Sub New()
            fileController = New DefaultFileController()
        End Sub

        Public Function CanHandleCommand(ByVal command As ControllerCommand) As Boolean
            Select Case command
                Case ControllerCommand.Open
                    Return Not InProcess
                Case ControllerCommand.Save
                    Return Not InProcess AndAlso Sessions.Count <> 0
                Case ControllerCommand.SaveAs
                    Return Not InProcess AndAlso FileExists
                Case ControllerCommand.Start
                    Return Not InProcess
                Case ControllerCommand.Stop
                    Return InProcess
                Case ControllerCommand.Delete
                    Return Not InProcess AndAlso Sessions.Count <> 0
            End Select

            Return False
        End Function

        Public Sub Dispose() Implements IDisposable.Dispose
            [Stop]()
            If CanHandleCommand(ControllerCommand.Save) Then
                Save()
            ElseIf CanHandleCommand(ControllerCommand.SaveAs) Then
                SaveAs()
            End If
        End Sub

        Public Sub [Stop]()
            If CanHandleCommand(ControllerCommand.Stop) Then
                Try
                    Dim processor As TraceDataProcessor = New TraceDataProcessor(runningSession)
                    OnStop.Invoke(processor)
                    Update(processor)
                Finally
                    OnStop = Nothing
                    If File.Exists(path) Then File.Delete(path)
                    path = ""
                    InProcess = False
                End Try
            End If
        End Sub

        Public Function GetBenchmarks(ByVal session As Integer) As List(Of BenchmarkItem)
            Return Benchmarks(session)
        End Function

        Public Function GetTraceEvents(ByVal session As Integer) As List(Of TraceItem)
            Return TraceItems(session).Values.SelectMany(Function(x) x).ToList()
        End Function

        Public Function GetTraceEvents(ByVal session As Integer, ByVal current As BenchmarkItem) As List(Of TraceItem)
            Dim length As Integer = current.End - current.Start
            Return TraceItems(session)(current.ThreadId).GetRange(current.Start, length)
        End Function

        Private Sub WorkWithFile(ByVal command As ControllerCommand, ByVal process As Action)
            If CanHandleCommand(command) Then
                Try
                    process()
                Catch e As Exception
                    RaiseEvent ThrowException(e)
                End Try
            End If
        End Sub

        Public Sub Open(ByVal Optional openName As String = "")
            WorkWithFile(ControllerCommand.Open, Sub()
                Dim fileName As String
                If fileController.TryOpenFile(fileName, openName) Then
                    Clear()
                    Me.fileName = fileName
                    Using stream = New MemoryStream(File.ReadAllBytes(fileName))
                        Dim [set] = BenchmarkHelper.LoadBenchmarkSet(stream)
                        Load([set])
                    End Using
                End If
            End Sub)
        End Sub

        Public Sub Save()
            WorkWithFile(ControllerCommand.Save, Sub()
                If FileExists OrElse fileController.TrySaveFile(fileName) Then
                    SaveBenchmarks(fileName)
                Else
                    SaveAs()
                End If
            End Sub)
        End Sub

        Public Sub SaveAs(ByVal Optional saveName As String = "")
            WorkWithFile(ControllerCommand.SaveAs, Sub()
                Dim fileName As String
                If fileController.TrySaveFile(fileName, If(String.IsNullOrEmpty(Me.fileName), saveName, IO.Path.GetFileName(Me.fileName))) Then
                    SaveBenchmarks(fileName)
                    Me.fileName = fileName
                End If
            End Sub)
        End Sub

        Private Sub SaveBenchmarks(ByVal fileName As String)
            Using stream = File.Create(fileName)
                BenchmarkHelper.SaveBenchmarkItems(Benchmarks, TraceItems, Sessions, stream)
            End Using
        End Sub

        Private Sub Load(ByVal [set] As BenchmarkItemSet)
            Loading(Benchmarks, [set].GetBenchmarks())
            LoadingTraceItems(TraceItems, [set].GetTraceItems())
            Sessions.AddRange([set].Sessions)
            runningSession = 0
            Dim number As Integer = Sessions.Count
            For i As Integer = 0 To number - 1
                Dim sessionId As Integer = Sessions(i).ID
                runningSession = Math.Max(sessionId, runningSession)
            Next

            RaiseEvent SessionsChanged()
        End Sub

        Private Sub Loading(Of T As ItemBase)(ByVal current As Dictionary(Of Integer, List(Of T)), ByVal [get] As Dictionary(Of Integer, List(Of T)))
            For Each pair As KeyValuePair(Of Integer, List(Of T)) In [get]
                Dim key As Integer = pair.Key
                current.Add(key, New List(Of T)())
                current(key).AddRange(pair.Value)
            Next
        End Sub

        Private Sub LoadingTraceItems(ByVal current As Dictionary(Of Integer, Dictionary(Of Integer, List(Of TraceItem))), ByVal [get] As Dictionary(Of Integer, Dictionary(Of Integer, List(Of TraceItem))))
            For Each session As KeyValuePair(Of Integer, Dictionary(Of Integer, List(Of TraceItem))) In [get]
                Dim key As Integer = session.Key
                current.Add(key, New Dictionary(Of Integer, List(Of TraceItem))())
                Loading(current(key), session.Value)
            Next
        End Sub

        Private Sub Clear()
            fileName = ""
            TraceItems.Clear()
            Benchmarks.Clear()
            Sessions.Clear()
            runningSession = 0
            RaiseEvent SessionsChanged()
        End Sub

        Public Sub Delete(ByVal item As SessionItem)
            If item Is Nothing Then Return
            Dim id As Integer = item.SessionId
            TraceItems.Remove(id)
            Benchmarks.Remove(id)
            For i As Integer = 0 To Sessions.Count - 1
                If Sessions(i).ID = id Then Sessions.RemoveAt(i)
            Next

            RaiseEvent SessionsChanged()
            runningSession = Enumerable.DefaultIfEmpty(Of Integer)(Enumerable.Select(Of SessionItem, Global.System.Int32)(Sessions, CType(Function(s) CInt(s.ID), Func(Of SessionItem, Integer))), CInt(0)).Max()
        End Sub

        Public Sub Start()
            InProcess = True
            Sessions.Add(New SessionItem(Interlocked.Increment(runningSession)))
            Benchmarks.Add(runningSession, New List(Of BenchmarkItem)())
            TraceItems.Add(runningSession, New Dictionary(Of Integer, List(Of TraceItem))())
            RaiseEvent SessionsChanged()
            OnStop = ReadBenchmarks()
        End Sub

        Private Function ReadBenchmarks() As Action(Of TraceDataProcessor)
            path = IO.Path.Combine(AppDomain.CurrentDomain.BaseDirectory, $"DX-Diagnostic.etl")
            If File.Exists(path) Then File.Delete(path)
            If Environment.OSVersion.Platform = PlatformID.Unix Then
                Return ReadBenchmarksLinux()
            Else
                Return ReadBenchmarksWindows()
            End If
        End Function

        Private Function ReadBenchmarksWindows() As Action(Of TraceDataProcessor)
            Dim session = New TraceEventSession(place, path)
            Dim id As Guid = TraceEventProviders.GetEventSourceGuidFromName(telemetry)
            session.EnableProvider(id)
            Thread.Sleep(100)
            Return Sub(processor)
                session.DisableProvider(id)
                session.Stop(True)
                Using source = New ETWTraceEventSource(path)
                     ''' Cannot convert LocalFunctionStatementSyntax, CONVERSION ERROR: Conversion for LocalFunctionStatement not implemented, please report this issue in 'void action(Microsoft.Diagn...' at character 12292
''' 
''' 
''' Input:
'''                     void action(Microsoft.Diagnostics.Tracing.TraceEvent data) => processor.Process(new DiagnosticTool.LinuxTools.DXTraceEvent(data));
''' 
'''  AddHandler source.Dynamic.All, Me.action
                    source.Process()
                    RemoveHandler source.Dynamic.All, Me.action
                End Using

                session.Dispose()
            End Sub
        End Function

        Private Sub ExecuteBash(ByVal cmd As String)
            Dim escapedArgs = cmd.Replace("""", "\""")
            Dim process = New Process With {.StartInfo = New ProcessStartInfo With {.FileName = "bash", .Arguments = $"-c ""{escapedArgs}""", .UseShellExecute = False, .CreateNoWindow = True}, .EnableRaisingEvents = True}
            process.Start()
            process.WaitForExit()
            process.Kill()
        End Sub

        Private Function ReadBenchmarksLinux() As Action(Of TraceDataProcessor)
            ExecuteBash(LinuxCommands.StartSession)
            Return Sub(processor)
                ExecuteBash(LinuxCommands.StopSession(path))
                Dim parser As DXTraceEventParser = New DXTraceEventParser()
                parser.Process(Sub(data) processor.Process(data), path)
            End Sub
        End Function

        Private Sub Update(ByVal processor As TraceDataProcessor)
            RaiseEvent StartProcessing()
            Try
                Benchmarks(runningSession) = New List(Of BenchmarkItem)(processor.Benchmarks)
                TraceItems(runningSession) = New Dictionary(Of Integer, List(Of TraceItem))(processor.TraceItems)
            Finally
                RaiseEvent EndProcessing()
                RaiseEvent SessionsChanged()
            End Try
        End Sub
    End Class
End Namespace

File: CS/DiagnosticTool/DignosticController/DiagnosticController.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Threading;
using DashboardDiagnosticTool.Data;
using DiagnosticTool.LinuxTools;
using Microsoft.Diagnostics.Tracing;
using Microsoft.Diagnostics.Tracing.Session;

namespace DashboardDiagnosticTool {
    public class DiagnosticController : IDisposable {
        string fileName = "";
        int runningSession = 0;
        string path;
        Action<TraceDataProcessor> OnStop;

        readonly IFileController fileController;
        readonly string place = "Session";
        readonly string telemetry = "DX-Dashboard-Telemetry";
        readonly Dictionary<int, Dictionary<int, List<TraceItem>>> traceItems = new Dictionary<int, Dictionary<int, List<TraceItem>>>();
        readonly Dictionary<int, List<BenchmarkItem>> benchmarks = new Dictionary<int, List<BenchmarkItem>>();
        readonly List<SessionItem> sessions = new List<SessionItem>();

        public bool InProcess = false;
        public bool FileExists => File.Exists(fileName);
        public Dictionary<int, Dictionary<int, List<TraceItem>>> TraceItems { get => traceItems; }
        public Dictionary<int, List<BenchmarkItem>> Benchmarks { get => benchmarks; }
        public List<SessionItem> Sessions { get => sessions; }

        public event Action<Exception> ThrowException;
        public event Action SessionsChanged;
        public event Action StartProcessing;
        public event Action EndProcessing;

        public DiagnosticController(IFileController fileController) {
            this.fileController = fileController;
        }

        public DiagnosticController() {
            this.fileController = new DefaultFileController();
        }

        public bool CanHandleCommand(ControllerCommand command) {
            switch(command) {
                case ControllerCommand.Open:
                    return !InProcess;
                case ControllerCommand.Save:
                    return !InProcess && Sessions.Count != 0;
                case ControllerCommand.SaveAs:
                    return !InProcess && FileExists;
                case ControllerCommand.Start:
                    return !InProcess;
                case ControllerCommand.Stop:
                    return InProcess;
                case ControllerCommand.Delete:
                    return !InProcess && Sessions.Count != 0;
            }
            return false;
        }
        public void Dispose() {
            Stop();
            if(CanHandleCommand(ControllerCommand.Save))
                Save();
            else if(CanHandleCommand(ControllerCommand.SaveAs))
                SaveAs();
        }
        public void Stop() {
            if(CanHandleCommand(ControllerCommand.Stop)) {
                try {
                    TraceDataProcessor processor = new TraceDataProcessor(runningSession);
                    OnStop.Invoke(processor);
                    Update(processor);
                } finally {
                    OnStop = null;
                    if(File.Exists(path))
                        File.Delete(path);
                    path = "";
                    InProcess = false;
                }
            }
        }
        public List<BenchmarkItem> GetBenchmarks(int session) {
            return Benchmarks[session];
        }
        public List<TraceItem> GetTraceEvents(int session) {
            return TraceItems[session].Values.SelectMany(x => x).ToList();
        }
        public List<TraceItem> GetTraceEvents(int session, BenchmarkItem current) {
            int length = current.End - current.Start;
            return TraceItems[session][current.ThreadId]
                .GetRange(current.Start, length);
        }
        void WorkWithFile(ControllerCommand command, Action process) {
            if(CanHandleCommand(command))
                try {
                    process();
                } catch(Exception e) {
                    ThrowException.Invoke(e);
                }
        }
        public void Open(string openName = "") {
            WorkWithFile(ControllerCommand.Open, () => {
                string fileName;
                if(fileController.TryOpenFile(out fileName, openName)) {
                    Clear();
                    this.fileName = fileName;
                    using(var stream = new MemoryStream(File.ReadAllBytes(fileName))) {
                        var set = BenchmarkHelper.LoadBenchmarkSet(stream);
                        Load(set);
                    }
                }
            });
        }
        public void Save() {
            WorkWithFile(ControllerCommand.Save, () => {
                if(FileExists || fileController.TrySaveFile(out fileName))
                    SaveBenchmarks(fileName);
                else SaveAs();
            });
        }
        public void SaveAs(string saveName = "") {
            WorkWithFile(ControllerCommand.SaveAs, () => {
                string fileName;
                if(fileController.TrySaveFile(out fileName,
                    string.IsNullOrEmpty(this.fileName) ? saveName: Path.GetFileName(this.fileName))) {
                    SaveBenchmarks(fileName);
                    this.fileName = fileName;
                }
            });
        }
        void SaveBenchmarks(string fileName) {
            using(var stream = File.Create(fileName)) {
                BenchmarkHelper.SaveBenchmarkItems(Benchmarks, TraceItems, Sessions, stream);
            }
        }
        void Load(BenchmarkItemSet set) {
            Loading(Benchmarks, set.GetBenchmarks());
            LoadingTraceItems(TraceItems, set.GetTraceItems());
            Sessions.AddRange(set.Sessions);
            runningSession = 0;
            int number = Sessions.Count;
            for(int i = 0; i < number; i++) {
                int sessionId = Sessions[i].ID;
                runningSession = Math.Max(sessionId, runningSession);
            }
            SessionsChanged?.Invoke();
        }
        void Loading<T>(Dictionary<int, List<T>> current, Dictionary<int, List<T>> get) where T : ItemBase {
            foreach(KeyValuePair<int, List<T>> pair in get) {
                int key = pair.Key;
                current.Add(key, new List<T>());
                current[key].AddRange(pair.Value);
            }
        }
        void LoadingTraceItems(Dictionary<int, Dictionary<int, List<TraceItem>>> current,
            Dictionary<int, Dictionary<int, List<TraceItem>>> get) {
            foreach(KeyValuePair<int, Dictionary<int, List<TraceItem>>> session in get) {
                int key = session.Key;
                current.Add(key, new Dictionary<int, List<TraceItem>>());
                Loading(current[key], session.Value);
            }
        }
        void Clear() {
            fileName = "";
            TraceItems.Clear();
            Benchmarks.Clear();
            Sessions.Clear();
            runningSession = 0;
            SessionsChanged?.Invoke();
        }
        public void Delete(SessionItem item) {
            if(item == null) return;
            int id = item.SessionId;
            TraceItems.Remove(id);
            Benchmarks.Remove(id);
            for(int i = 0; i < Sessions.Count; i++) {
                if(Sessions[i].ID == id)
                    Sessions.RemoveAt(i);
            }
            SessionsChanged?.Invoke();
            runningSession = Sessions.Select(s => s.ID).DefaultIfEmpty(0).Max();
        }
        public void Start() {
            InProcess = true;
            Sessions.Add(new SessionItem(++runningSession));
            Benchmarks.Add(runningSession, new List<BenchmarkItem>());
            TraceItems.Add(runningSession, new Dictionary<int, List<TraceItem>>());
            SessionsChanged?.Invoke();
            OnStop = ReadBenchmarks();
        }

        Action<TraceDataProcessor> ReadBenchmarks() {
            path = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, $"DX-Diagnostic.etl");
            if(File.Exists(path))
                File.Delete(path);
            if(Environment.OSVersion.Platform == PlatformID.Unix)
                return ReadBenchmarksLinux();
            else
                return ReadBenchmarksWindows();
        }
        Action<TraceDataProcessor> ReadBenchmarksWindows() {
            var session = new TraceEventSession(place, path);
            Guid id = TraceEventProviders.GetEventSourceGuidFromName(telemetry);
            session.EnableProvider(id);
            Thread.Sleep(100);
            return processor => {
                session.DisableProvider(id);
                session.Stop(true);
                using(var source = new ETWTraceEventSource(path)) {
                    void action(TraceEvent data) => processor.Process(new DXTraceEvent(data));
                    source.Dynamic.All += action;
                    source.Process();
                    source.Dynamic.All -= action;
                }
                session.Dispose();
            };
        }

        void ExecuteBash(string cmd) {
            var escapedArgs = cmd.Replace("\"", "\\\"");
            var process = new Process {
                StartInfo = new ProcessStartInfo {
                    FileName = "bash",
                    Arguments = $"-c \"{escapedArgs}\"",
                    UseShellExecute = false,
                    CreateNoWindow = true
                },
                EnableRaisingEvents = true
            };
            process.Start();
            process.WaitForExit();
            process.Kill();
        }

        Action<TraceDataProcessor> ReadBenchmarksLinux() {
            ExecuteBash(LinuxCommands.StartSession);
            return processor => {
                ExecuteBash(LinuxCommands.StopSession(path));
                DXTraceEventParser parser = new DXTraceEventParser();
                parser.Process(data => processor.Process(data), path);
            };
        }

        void Update(TraceDataProcessor processor) {
            StartProcessing?.Invoke();
            try {
                Benchmarks[runningSession] = new List<BenchmarkItem>(processor.Benchmarks);
                TraceItems[runningSession] = new Dictionary<int, List<TraceItem>>(processor.TraceItems);
            } finally {
                EndProcessing?.Invoke();
                SessionsChanged?.Invoke();
            }
        }
    }
}
